src/util/thread_allocator.cpp

```diff
@@ -0,0 +1,494 @@
+            thread_local_allocator* allocator =
+                &thread_local_allocator::get_tls_allocator();
+            allocator->process_pending_requests();
+
+            std::size_t i = block_chain[size / BLOCK_ALIGNMENT];
+
+            blk = allocator->chain[i];
+            if (HPX_LIKELY(blk != nullptr))
+            {
+                allocator->chain[i] = blk->next_free;
+                blk->set_size(static_cast<std::int16_t>(size));
+            }
+            else
+            {
+                detail::alloc_page* pg = reinterpret_cast<detail::alloc_page*>(
+                    std::malloc(sizeof(detail::alloc_page)));
+                if (nullptr == pg)
+                {
+                    return nullptr;
+                }
+                new (pg) detail::alloc_page(allocator);
+
+                allocator->pages = pg;
+
+                std::size_t block_size = chain_block_size[i];
+                std::size_t blocks =
+                    detail::alloc_page::page_size / block_size - 2;
+
+                // FIXME: move block construction into alloc_page constructor
+                allocator->chain[i] = pg->get()->next(block_size);
+
+                blk = allocator->chain[i];
+                new (blk) detail::alloc_block(pg);
+
+                do
+                {
+                    detail::alloc_block* next = blk->next(block_size);
+                    new (next) detail::alloc_block(pg);
+
+                    blk->next_free = next;
+                    blk = next;
+                } while (--blocks != 0);
+
+                blk->next_free = nullptr;
+
+                blk = pg->get();
+                new (blk)
+                    detail::alloc_block(pg, static_cast<std::int16_t>(size));
+
+                blk->next_free = nullptr;
+            }
+
+            ++blk->page->allocated_blocks;
+        }
+        else
+        {
+            blk = reinterpret_cast<detail::alloc_block*>(std::malloc(size));
+            if (blk == nullptr)
+            {
+                return nullptr;
+            }
+            new (blk) detail::alloc_block(nullptr, MAX_BLOCK_SIZE + 1);
+        }
+
+        return reinterpret_cast<detail::alloc_block_header*>(blk) + 1;
+    }
+
+    void thread_free(void* addr)
+    {
+        if (addr == nullptr)
+        {
+            return;     // ignore nullptr arguments
+        }
+
+        detail::alloc_block* blk = static_cast<detail::alloc_block*>(
+            reinterpret_cast<detail::alloc_block_header*>(addr) - 1);
+
+        if (blk->get_size() > MAX_BLOCK_SIZE)
+        {
+            blk->detail::alloc_block::~alloc_block();
+            std::free(blk);
+        }
+        else
+        {
+            detail::alloc_page* page = blk->get_ptr();
+            thread_local_allocator* allocator = page->allocator;
+            thread_local_allocator* current =
+                &thread_local_allocator::get_tls_allocator();
+
+            if (allocator != current)
+            {
```

Comment:
We need to add a `hpx::util::ignore_all_while_checking il;` here. When destructors are run, we might hold some locks.

Reply:
Fair point, thanks. Done.